ddtrace/contrib/aws_lambda/patch.py

```diff
@@ -124,16 +124,35 @@ def _get_handler_and_module():
     if path is None:
         from datadog_lambda.wrapper import datadog_lambda_wrapper
 
-        handler = getattr(datadog_lambda_wrapper, "__call__")
+        wrapper_module = datadog_lambda_wrapper
+        wrapper_handler = getattr(datadog_lambda_wrapper, "__call__")
 
-        return handler, datadog_lambda_wrapper, _datadog_instrumentation
+        return wrapper_handler, wrapper_module, _datadog_instrumentation
     else:
         parts = path.rsplit(".", 1)
         (mod_name, handler_name) = parts
         modified_mod_name = _modify_module_name(mod_name)
         handler_module = import_module(modified_mod_name)
         handler = getattr(handler_module, handler_name)
-        return handler, handler_module, _datadog_instrumentation
+
+        if callable(handler):
+            # handler is a function
+            if not isinstance(handler, type) and hasattr(handler, "__code__"):
```

Comment:
Would be worthwhile to add a comment here to indicate that `hasattr(handler, "__code__")` is a "best effort" attempt to detect whether this is a function or a class instance.

Technically adding a `__code__` property to the class is valid (though really strange), so a comment would make it more explicit that we're not accounting for such a use case. Unfortunately I haven't been able to find a way to distinguish between callable class and a function without "reflection" hacks like defining a function, capturing it's type and comparing against that, but even that doesn't cover all cases.

Reply:
Ah, you're right! 

Technically, it's valid yeah. I'll add a comment here to address it. 
I haven't been successful to find an example to distinguish it either. 
These won't cover all cases. But we are expecting to go to back to normal and remove the monkey patch once our tracing is done only in this repo.

Reply:
I was thinking, for this use case, I can just add another condition to the statement, like this:
What do you think? 

```python
class_name = type(handler).__name__
is_function = not isinstance(handler, type) and hasattr(handler, "__code__") and class_name == "function"
if is_function:
    return handler, handler_module, _datadog_instrumentation

# class instance with __code__ defined will ignore it, and use __call__ instead here
...
```

Reply:
That's an option, though it doesn't cover all cases. For example technically (and again, mostly just trying to see how I can break this), `print` is a valid lambda handler, but it's type is `<class 'builtin_function_or_method'>` and would thus miss this check. 

Something like this may be a bit more robust:
```
from types import (
    BuiltinFunctionType,
    BuiltinMethodType,
    ClassMethodDescriptorType,
    FunctionType,
    LambdaType,
    MethodDescriptorType,
    MethodWrapperType,
    WrapperDescriptorType,
)

AnyFunction = (
    BuiltinFunctionType,
    BuiltinMethodType,
    ClassMethodDescriptorType,
    FunctionType,
    LambdaType,
    MethodDescriptorType,
    MethodWrapperType,
    WrapperDescriptorType,
)

def is_function(callable: Any) -> bool:
    return isinstance(callable, AnyFunction)
```
But that's assuming you want to cover all such cases, which may not be the case. If not, then a comment indicating that it's a best effort attempt to check against user defined functions.

Reply:
You're right. I'll leave it as a best effort.
I'll add a comment and a warning log for this, I'll also make sure we only warn on cold starts, in order to avoid logging on every invocation. 

Thank you so much @michaeldimchuk